src/kv/kv.h

```diff
@@ -1495,80 +1498,106 @@ namespace kv
-      // same store will result in a store version mismatch and
-      // deserialisation will then fail.
+      frame::FlatbufferDeserialiser fbd(data.data());
+      auto frames = fbd.get_frames();
+      Version v;
+      OrderedViews<S, D> views;
+      bool first_serialiser = true;
       auto e = get_encryptor();
-      D d(
-        e,
-        public_only ? kv::SecurityDomain::PUBLIC :
-                      std::optional<kv::SecurityDomain>());
-      if (!d.init(data))
+
+      for (auto& [frame, size] : frames)
       {
-        LOG_FAIL_FMT("Initialisation of deserialise object failed");
-        return DeserialiseSuccess::FAILED;
-      }
+        if (size > 0)
+        {
+          // This will return FAILED if the serialised transaction is being
+          // applied out of order.
+          // Processing transactions locally and also deserialising to the
+          // same store will result in a store version mismatch and
+          // deserialisation will then fail.
+          D d(
+            e,
+            public_only ? kv::SecurityDomain::PUBLIC :
+                          std::optional<kv::SecurityDomain>());
+
+          if (!d.init(frame, size))
+          {
+            LOG_FAIL_FMT("Initialisation of deserialise object failed");
+            return DeserialiseSuccess::FAILED;
+          }
 
-      Version v = d.template deserialise_version<Version>();
-      LOG_DEBUG_FMT("Deserialising {}", v);
+          if (first_serialiser)
+          {
+            first_serialiser = false;
+            v = d.template deserialise_version<Version>();
+            // Throw away any local commits that have not propagated via the
+            // consensus.
+            rollback(v - 1);
+
+            // Make sure this is the next transaction.
+            auto cv = current_version();
+            if (cv != (v - 1))
+            {
+              LOG_FAIL_FMT(
+                "Tried to deserialise {} but current_version is {}", v, cv);
+              return DeserialiseSuccess::FAILED;
+            }
+          }
+          else
+          {
+            Version v_ = d.template deserialise_version<Version>();
+            LOG_DEBUG_FMT("Deserialising {}", v_);
+            if (v != v_)
+            {
+              LOG_FAIL_FMT("Deserialisers versions do not match {} {}", v, v_);
+              return DeserialiseSuccess::FAILED;
+            }
+          }
 
-      // Throw away any local commits that have not propagated via the
-      // consensus.
-      rollback(v - 1);
+          // Deserialised transactions express read dependencies as versions,
+          // rather than with the actual value read. As a result, they don't
+          // need snapshot isolation on the map state, and so do not need to
+          // lock all the maps before creating the transaction.
+          std::unordered_set<std::string> present;
+          std::lock_guard<SpinLock> mguard(maps_lock);
```

Comment:
We lose the atomicity of deserialise if we do this, and that seems unfortunate (ie. a tx can observe a Tx made of only one frame!)

Reply:
Very true! I have changed it so that we find the first frame that has data, create the first deserlaliser, get the version, rollback, and _then_ acquire the lock which will be held while we are deserializing everything else, committing, etc like before.